### PR TITLE
[SPARK-48260][SQL] Disable output committer coordination in one test of ParquetIOSuite

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

Recently I noticed a test from `ParquetIOSuite` being flaky: `SPARK-7837 Do not close output writer twice when commitTask() fails`

It turns out to be a race condition. The test injects error to the task committing step, and the job may fail in two ways:
1. The task got the driver's permission to commit the task, but the committing failed and thus the task failed. This will trigger a stage failure as it means possible data duplication, see https://github.com/apache/spark/pull/36564
2. In test we disable task retry, so `TaskSetManager` will abort the stage.

Both these two failures are done by sending an event to `DAGScheduler`, so the final job failure depends on which event gets processed first. This is not a big deal, but that test in `ParquetIOSuite` checks the error class. This PR fixes the flaky test by disabling output committer coordination. Necessary changes are added to allow the disablement per query.

### Why are the changes needed?

fix flaky test

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

N/A

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @gengliangwang @viirya

**Comment by viirya** [comment]
> The task got the driver's permission to commit the task, but the committing failed and thus the task failed. This will trigger a stage failure as it means possible data duplication, see https://github.com/apache/spark/pull/36564

I read https://github.com/apache/spark/pull/36564, it seems to handle the case that a commit is successful but the task is failed which means data duplication is possible.

But here the task and the commit are both failed. It should not cause data duplication. Does it also trigger a stage failure?

**Comment by cloud-fan** [comment]
@viirya I think https://github.com/apache/spark/pull/36564 's PR description was very clear about the details. Driver never knows if the task commit is successful or not. It only knows: 1) if the "ask for commit" request is approved or not. 2) if the task completes successfully or not.

The flaky test triggers it: 1) it injects failure to `commitTask`, which happens after the task receives driver's permission to commit. 2) the task fails eventually.

**Comment by viirya** [comment]
> In this pr, we do below since:
> When commit task success, executor send an CommitOutputSuccess message to outputCommitCoordinator.
> When outputCommitCoordinator handle taskComplete, if task failed but commit success, means data duplicate will happen, we should failed to job.

@cloud-fan Thanks for explaining it. It sounds reasonable. Although from what I read from the above description in https://github.com/apache/spark/pull/36564, it seems when outputCommitCoordinator handles taskComplete, it will check if the task is failed but the commit is successful, then decide if data duplication is happened or not.

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
> The task got the driver's permission to commit the task, but the committing failed and thus the task failed. This will trigger a stage failure as it means possible data duplication

This is more correct. The `OutputCommitCoordinator` only knows that a commit is authorized (the task got the driver's permission). But in #36565, it is not distinct it from a commit success. It'd be better to update the reason string in `OutputCommitCoordinator`, although not related to this PR.

File: core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala [hunk 1]
```diff
@@ -276,8 +276,14 @@ class HadoopMapReduceCommitProtocol(
   override def commitTask(taskContext: TaskAttemptContext): TaskCommitMessage = {
     val attemptId = taskContext.getTaskAttemptID
     logTrace(s"Commit task ${attemptId}")
+    val disableCommitCoordination =
```
[the remark this answers is not in the document]
Comment (gengliangwang) [reply core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala:279]: Also let's add a comment about why we need to disable it in test

File: core/src/main/scala/org/apache/spark/mapred/SparkHadoopMapRedUtil.scala [hunk 2]
```diff
@@ -42,7 +42,8 @@ object SparkHadoopMapRedUtil extends Logging {
       committer: MapReduceOutputCommitter,
       mrTaskContext: MapReduceTaskAttemptContext,
       jobId: Int,
-      splitId: Int): Unit = {
+      splitId: Int,
+      disableCommitCoordination: Boolean): Unit = {
```
Comment (gengliangwang) [inline core/src/main/scala/org/apache/spark/mapred/SparkHadoopMapRedUtil.scala:46]: nit: shall we rename as `disableCommitCoordinationInTest` as well?

**Comment by gengliangwang** [comment]
@cloud-fan I created https://github.com/apache/spark/pull/46562 for this, which avoids changing the production code to fix the flaky test.

**Comment by cloud-fan** [comment]
closing in favor of https://github.com/apache/spark/pull/46562

**Pull request closed without merging**